README
IPTables-libiptc version 0.51
=============================

This package provides a Perl interface to the netfilter/iptables
C-library libiptc.

Advantages of this module: Many rule changes can be done very
fast. Several rule changes is committed atomically.

This module is heavily inspired by the CPAN module IPTables-IPv4.  The
CPAN module IPTables-IPv4 could not be used because it has not been
kept up-to-date, with the newest iptables extensions.  This is a
result of the module design, as it contains every extension and thus
needs to port them individually.

This package has another approach, it links with the systems
libiptc.so and libxtables.so library and depend on dynamic loading of
iptables extensions available on the system.

This design limits us to only export the libiptc chain manipulation
functions.  All rule manipulations are done through the iptables.c
"do_command" function.  As iptables.c is not made as a library, the
package unfortunately needs to maintain/contain this C file. This
unfortunately also means that we need to keep up-to-date with
iptables.c API changes (but significantly less work that keeping in
sync with every extension module).


DEPENDENCIES

This module requires these other modules and libraries:

 iptables
 xtables.h
 libiptc.so
 libxtables.so

The iptables version installed on your system is auto-detected during
generation of the Makefile.PL file.  This sets the IPTABLES_VERSION
variable as it needs to match the .so extentions available on you
system.

This release focus on supporting iptables above version 1.4.3.2
and up-to iptables version 1.4.10.

The default fallback IPTABLES_VERSION is currently set to version 1.4.4.

If you need support for iptables versions 1.4.1 and 1.4.2, we
recommend using version 0.18 of this module IPTables::libiptc.


BUGS

Significant API changes when into 1.4.3, and thus older versions cannot
be kept backward binary compatible with this release.


INSTALLATION

Notice this module depends on iptables dynamic loadable libraries,
thus it needs to know the path to them (default assumes they are
located in /lib/xtables).  The iptables path for dynamic loading
(where the lib{ip,x}t_*.so files are located) can be altered by setting
the variable XTABLES_LIBDIR while calling make.  This can also be
changed later at runtime setting the environment variable
XTABLES_LIBDIR.

To install this module type the following:

   perl Makefile.PL PREFIX=/usr/local/
   make XTABLES_LIBDIR=/usr/local/libexec/xtables
   sudo make test
   make install PREFIX=/usr/local/stow/IPTables-libiptc-0.51

   cd /usr/local/stow/
   stow IPTables-libiptc-0.51


COPYRIGHT AND LICENCE

Copyright (C) 2006-2011 by Jesper Dangaard Brouer <hawk@cpan.org>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.


Changes
Revision history for Perl extension IPTables::libiptc.

Recent changes

0.51    Wed Jun 29 11:37:42 EDT 2011

	- Support for iptables v1.4.3 until v1.4.10.

	- This is a significant change release, in-order to adapt to
          iptables ABI/API changes in release 1.4.3, which introduced
          dynamic shared libs.

	- Using the dynamic shared libs libiptc.so and libxtables.so,
          available on target system, instead of maintaining this code
          locally.

	- This version requires that the system provides a correct
          "xtables.h" header file, which corrosponds to the systems
          libxtables.so shared library.

	- To keep the iptables.c file easy maintainable, several
          original versions are included in this package.  The choosen
          versions corrospond to the API/ABI changes in libxtables.


0.18	Mon Oct 26 14:23:51 CET 2009

	- This version is compatible with iptables 1.4.1 and 1.4.2.

	- Significant changes when into 1.4.3, which thus cannot be
	  kept binary compatible with this release.  Expect to see the
	  following error "undefined symbol: xt_params" if using this
	  release anyhow.

	- The segfault issues when comparing the IPTABLES_VERSION
	  variable is handled, and is based upon the iptables detected
	  by the Makefile.PL script.  And some compile time hacks to
	  include/xtables.h


0.17	Fri Oct 23 17:21:23 CEST 2009

	- This version is only compatible with iptables 1.4.1.

	- Its not compatible with 1.4.2 and 1.4.3, because there are
	  some segfault issues when comparing the IPTABLES_VERSION
	  variable (in the dynamic loadable modules) between iptables
	  versions, 1.4.1 and 1.4.2, and also between 1.4.2 and 1.4.3.
	  It has been fixed from version 1.4.3.2 and beyond (see
	  iptables upstream git commit c4edfa63eda06).

	- This is the first version to support iptables 1.4.x, more
	  will be created shortly, for supporting 1.4.2 and 1.4.3.

	- This version also removes the work around for exit()
	  problem, done in release 0.13.  Its still a problem, but
	  its too intrusive to the iptables.c code.


0.15  Fri Jul 10 11:58:47 CEST 2009

	- Problem with list_rules_IPs() on 64-bit platforms.
	  Fixed by Colm Dougan <colm.dougan(at)replify(dot)com>


0.14  Thu Apr 16 11:37:16 CEST 2009

	- Auto-detect iptables version, but this release is only
	  compatible with all iptables 1.3.x versions.

	- This version is up-to-date with iptables 1.3.8.


0.13  Wed Apr 15 14:59:42 CEST 2009

	- Work around for exit() problem, by Alexandre Simon.
	  The C-code in iptables.c calls exit() on all errors, this
	  results in the Perl program dying.  This is now avoided by a
	  setjmp/longjmp hack by Alexandre(dot)Simon(at)ciril(dot)fr.


0.12  Tue Apr 14 15:11:32 CEST 2009

	- This version is compatible with iptables 1.3.6.

	- Use iptables.c command interface from 1.3.6.

	- Backport one of my own mainline bugfixes to chain rename
	  (upstream commit 64ff47cd, Author: Jesper Dangaard Brouer).

	- Backport bugfix "Fix /etc/network usage"
	  (upstream commit 267a5700, Author: Pablo Neira Ayuso).

	- Backport bugfix for rename command line option -E.
	  (upstream commit 2ac1fada, Author: Krzysztof Piotr Oledzki).


0.11  Thu Jul 10 10:22:39 CEST 2008

	- Skip test if not run as root user.  As I cannot make CPAN
	  tester run 'make test' as root, I simply skip these tests.
	  Credit goes to Brian D Foy, for the idea.

0.10  Tue Jul  1 13:38:56 CEST 2008

	- Add a -fPIC option to compile code as "position independent code".
	  As suggested by Matthew "Cheetah" Gabeler-Lee.  This also
	  caused problems for CPAN tester reports trying to compile it
	  on a 64-bit platform.

	- The local libiptc has been updated, to keep it up-to-date
	  with iptables libiptc 1.4.1, except for the xtables.h
	  include. But the iptables.c command interface is still from 1.3.4.

0.09  Tue Jun 17 16:06:33 CEST 2008
	- Fix output for unspecified proto was "0" change back to
	  original behaviour "all".

	- iptables-blocking: link explicit with local library
	  libiptc.a.  Install in sbin/.

0.08  Mon Jun 16 17:11:51 CEST 2008
	- New implementaion of the the "init" scalability fix, which
	  is 64-bit compatible.  This implementation is based on binary
	  search (time complexity has gone from O(n^2) to O(log(n)).

	- Revert the not accepted "init" scalability fix.

0.07  Mon Jun 16 07:08:33 CEST 2008
	- Includes the "init" scalability fix, NOT accepted upstream,
          because its not 64-bit compatible.  The initial parsing of
          the ruleset, during pull-out from kernel, is suboptimal.
	  Its gets slower when the number of chains and jump rules
	  increase.

	http://thread.gmane.org/gmane.comp.security.firewalls.netfilter.devel/21458

0.06  Fri Jun 13 17:13:21 CEST 2008
	- Matthew "Cheetah" Gabeler-Lee <cheetah(at)fastcat.org>:
		Make it compile even if iptables libiptc is not on the
		system, by linking with local library object libiptc in
		library_iptc/libiptc.a.

	- libiptc: Integrated scalability fixes. Most of which has
	  been accepted upstream in 1.4.1.  Most significant
	  scalability fix is for chain name searching, used in
	  function is_chain() and iptables_do_command().

	- The local libiptc has been keept up-to-date with iptables
	  1.4.0. But the iptables.c command interface is still from 1.3.4.

	- iptables-blocking: Add an iptables command that does uses a
	  simple file lock. Can be used to replace the iptables
	  command to avoid race conditions between two iptables commands.


0.05  Tue Nov 20 12:47:11 2007
	- Bugfix: Include all files in tarball by updating MANIFEST.
	  Version 0.04 could not compile, because the tarball did not
	  contain all the relevant files (ups, didn't keep MANIFEST in
	  sync).

0.04  Thu Nov 15 16:01:16 2007
	- First CPAN release.
	  Primarily documentation updates.

0.03  Thu Jul 13 16:24:07 2006
	- Because libiptc.c contained some bugs, it has been
          necessary to include it the module and compile libiptc.a our

Makefile.PL
use 5.008004;
use ExtUtils::MakeMaker;

# TODO: Create some test which detect if xtables.h is installed on the
# target system.  Under Debian this file is located in software
# package: "iptables-dev".

{
        package MY;
        use File::Basename; #for dirname

	sub exec_iptables($) {
	    my $cmd = shift;
	    my $exitcode;
	    print "Detect iptables version via command: \"$cmd\"\n";

	    my $output = `$cmd 2>&1`;
	    if ($? == -1) {
		# Cannot execute the command
		print(" WARN: Cannot execute the command:$cmd (err:$!)\n");
		$exitcode=-1;
	    } else {
		# The exit code is in the high byte of the 16-bit status word
		$exitcode = $? >> 8;
	    }
	    return ($exitcode, $output);
	}

	sub which_xtables_libdir($) {
	    # Find the xtables libdir based upon path of iptables binary
	    my $cmd = shift;
	    my $exitcode;
	    print "Detect the xtables libdir path of: \"$cmd\"\n";

	    my $output = `which $cmd 2>&1`;
	    my $xtables_libdir = "/lib/xtables";

	    if ($? == -1) {
		# Cannot execute the command
		print(" WARN: Cannot execute the command:$cmd (err:$!)\n");
		$exitcode=-1;
	    } else {
		# The exit code is in the high byte of the 16-bit status word
		$exitcode = $? >> 8;
		print(" WARN: error occured to run: which $cmd (err:$!)\n")
		    if ($exitcode > 0);
	    }
	    $output = dirname($output); # remove "iptables"
	    $output = dirname($output); # remove "sbin/"
	    if ($output =~ m/^\/$/ ) {
		# Only contains "/" means this was /sbin/iptables
		# and it has a special load path
		$xtables_libdir = "/lib/xtables";
	    } else {
		$xtables_libdir = "${output}/libexec/xtables"
	    }
	    print " - detected xtables libdir: \"$xtables_libdir\"\n";
	    return $xtables_libdir;
	}

	sub xtables_api_version($$$) {
	    #define XTABLES_API_VERSION(x,y,z)    (0x10000*(x) + 0x100*(y) + z)
	    my $major    = shift;
	    my $minor    = shift;
	    my $revision = shift;
	    my $version_code = 0x10000*($major) + 0x100*($minor) + $revision;
	    return $version_code;
	}

	sub detect_iptables_version($) {
	    my $default_version = shift;
	    my $ver = $default_version;
	    my $vercode;
	    my $cmd = "iptables -V";
	    my $xtables_libdir = "/lib/xtables";
	    my $exitcode;

	    ($exitcode, $output) = exec_iptables("iptables -V");
	    if ($exitcode == -1) {
		($exitcode, $output) = exec_iptables("/usr/local/sbin/iptables -V");
		if ($exitcode != -1) {
		    $xtables_libdir = "/usr/local/libexec/xtables";
		}
	    } else {
		# iptables found in the path, but where?  We need to
		# know which is the correct $xtables_libdir to use.
		$xtables_libdir = which_xtables_libdir("iptables");
	    }
	    ($exitcode, $output) = exec_iptables("/sbin/iptables -V")
		if ($exitcode == -1);

	    if ($exitcode>0) {
		print STDERR ("WARN: Err running command:\"$cmd\" (err:$!)\n");
		print STDERR ("WARN: Cannot auto-detect iptables version\n");
	    } else {
		chomp $output;
		#print "Matching on output:[$output]\n";
		if ($output =~ /iptables v(\d+)\.(\d+)\.(\d+)\.?(\d?)/) {
		    my $major    = $1;
		    my $minor    = $2;
		    my $revision = $3;
		    my $extra    = $4;
		    $ver="$major.$minor.$revision";
		    if ( $extra =~ m/\d+/ ) {
			$ver = $ver . "." . ${extra};
		    }
		    $vercode=xtables_api_version($major,$minor,$revision);
		    print "Detected iptables version: $ver (vercode:$vercode)\n";
		    if ($major > 1) {
			print "Too high major version, fallback\n";
			$ver = $default_version;
		    }

		    if ($minor > 4) {
			print "Module only supports version 1.4.x, fallback\n";
			$ver = $default_version
		    }

		    if ($minor < 4) {
			print "Version below 1.4.x, good luck\n";
			print "We recommend that you instead use version 0.15 of IPTables::libiptc\n";
			print "\n";
		    }

		    if ($minor == 4 && $revision < 3) {
			print "\n";
			print "-=-=-=- WARNING: Incompatible iptables version -=-=-=-";
			print "\n";
			print " This perl module release only supports iptables versions:\n";
			print "   v1.4.3.2 and above \n";
			print "\n";
			print "We recommend that you instead use version 0.18 of IPTables::libiptc\n";
			print "on your system, for supporting v1.4.1 and v1.4.2 \n";
			print "\n";
			print " Significant API changes when into 1.4.3, and thus older versions cannot\n";
			print " be kept backward binary compatible with this release.\n";
			print "\n";
			print " YOU HAVE BEEN WARNED!!!\n";
			print "\n";
		    }

		}
	    }
	    print("Using iptables version: $ver\n");
	    return ($ver, $vercode, $xtables_libdir);
	}

	sub post_initialize {
	    my $default_version = "1.4.4";
	    my ($ver, $vercode, $xtables_libdir) = detect_iptables_version($default_version);
	    #print "vercode: $vercode\n";
"
IPTABLES_VERSION:=$ver
IPTABLES_VERSION_CODE:=$vercode
XTABLES_LIBDIR:=$xtables_libdir
LOCAL_LIB_IPTC:=-Llibrary_iptc/
"
        }

        sub postamble {
"

# Not needed anymore
#library_iptc/libiptc.a: library_iptc/*.c
#	make -C library_iptc/ all IPTABLES_VERSION=\"\$(IPTABLES_VERSION)\" IPTABLES_VERSION_CODE=\"\$(IPTABLES_VERSION_CODE)\"

#iptables/iptables.o iptables/xtables.o: iptables/*.c library_iptc/libiptc.a
iptables/iptables.o iptables/xshared.o: iptables/*.c
	make -C iptables/ all PREFIX=\"\$(PREFIX)\" IPTABLES_VERSION=\"\$(IPTABLES_VERSION)\" XTABLES_LIBDIR=\"\$(XTABLES_LIBDIR)\" IPTABLES_VERSION_CODE=\"\$(IPTABLES_VERSION_CODE)\" EXTRALIBS=\"\$(EXTRALIBS)\"


install::
	make -C iptables/ install

clean::
	make -C iptables/ clean
#	make -C library_iptc/ clean
";
        }
}


# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'IPTables::libiptc',
    VERSION_FROM      => 'lib/IPTables/libiptc.pm', # finds $VERSION
    PREREQ_PM         => {}, # e.g., Module::Name => 1.1
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/IPTables/libiptc.pm', # retrieve abstract from module
       AUTHOR         => 'Jesper Dangaard Brouer <hawk@comx.dk>') : ()),
##  LIBS              => ['-ldl -lnsl -liptc'], # e.g., '-lm'
    # Notice, perl Makefile.PL will detect if ip4tc is available
    LIBS              => ['-ldl -lnsl -liptc -lip4tc -lxtables'], # e.g., '-lm'
##  LIBS              => ['-ldl -lnsl'], # e.g., '-lm'
    LDDLFLAGS         => '-shared -L$(PREFIX)/lib',
##  LDDLFLAGS         => '-shared $(LOCAL_LIB_IPTC) -L$(PREFIX)/lib',
    LDFLAGS           => '-L$(PREFIX)/lib',
    DEFINE            => '-g -DIPTABLES_VERSION=\"$(IPTABLES_VERSION)\" -DXTABLES_LIBDIR=\"\$(XTABLES_LIBDIR)\" -DIPTABLES_VERSION_CODE=\$(IPTABLES_VERSION_CODE)',
                         # e.g., '-DHAVE_SOMETHING'
    INC               => '-I/usr/local/include -I./include -I.',
                         # e.g., '-I. -I/usr/include/other'
    # Un-comment this if you add C files to link with later:
    # OBJECT            => '$(O_FILES)', # link all the C files too
    OBJECT            => '$(O_FILES) iptables/iptables.o iptables/xshared.o',
##  OBJECT            => '$(O_FILES) iptables/iptables.o iptables/xtables.o library_iptc/libiptc.a',
                      # link all the C files too
    PREFIX            => '/usr/local',
    TYPEMAPS          => ['libiptc.typemap'],
#   depend             => { 'iptables/iptables.o' => 'library_iptc/libiptc.a'}
#   depend             => { 'iptables/iptables.o' => 'library_iptc/libiptc.a'}
);
if  (eval {require ExtUtils::Constant; 1}) {
  # If you edit these definitions to change the constants used by this module,
  # you will need to use the generated const-c.inc and const-xs.inc
  # files to replace their "fallback" counterparts before distributing your
  # changes.
  my @names = (qw(IPT_MIN_ALIGN));
  ExtUtils::Constant::WriteConstants(
                                     NAME         => 'IPTables::libiptc',
                                     NAMES        => \@names,
                                     DEFAULT_TYPE => 'IV',
                                     C_FILE       => 'const-c.inc',
                                     XS_FILE      => 'const-xs.inc',
                                  );

}
else {
  use File::Copy;
  use File::Spec;
  foreach my $file ('const-c.inc', 'const-xs.inc') {
    my $fallback = File::Spec->catfile('fallback', $file);
    copy ($fallback, $file) or die "Can't copy $fallback to $file: $!";
  }
}
